index.bs

```diff
@@ -371,28 +403,164 @@ model]]:
+1. If |navigable|'s [=top-level traversable/bounce tracking record=]'s [=bounce tracking record/client bounce detection running=] is false:
+    1. Set |navigable|'s [=top-level traversable/bounce tracking record=] to a new
+        [=bounce tracking record=] with [=bounce tracking record/initial host=]
+        set to |initialHost|.
+1. Otherwise,
+    1. Let |isUserInitiated| denote whether this navigation was initiated by the user.
+    1. If |isUserInitiated|:
+        1. Run [=record stateful bounces for bounce tracking=].
+        1. Set |navigable|'s [=top-level traversable/bounce tracking record=] to a new [=bounce tracking record=] with
+            [=bounce tracking record/initial host=] set to |initialHost|.
+    1. Otherwise, add |initialHost| to |navigable|'s [=top-level traversable/bounce tracking record=]'s [=bounce tracking record/bounce set=].
+1. Set |navigable|'s [=top-level traversable/bounce tracking record=]'s [=bounce tracking record/client bounce detection running=] to false.
+
+Issue: TODO: Use a concrete spec signal to determine |isUserInitiated|.
+
+</div>
+
+<h5 id="bounce-tracking-mitigations-nav-end-monkey-patch">End Navigation Monkey
+Patch</h5>
+
+At the end of a navigation, fill the [=bounce tracking record/bounce set=].
+
+Insert the following steps in the <a spec="html">navigate</a> algorithm, under "Create navigation params by fetching", after Step 19.7,
+"Wait until either response is non-null...", or as soon as [=response/url list=] becomes available.
+
+1. Run [=process navigation end for bounce tracking=] given
+    <var ignore>navigable</var> and <var ignore>URLs</var>.
+
+<div algorithm>
+
+To <dfn>process navigation end for bounce tracking</dfn> given a [=navigable=]
+|navigable| and a [=response/url list=] |URLs|, perform the following steps:
+
+1. [=Assert=] that |navigable|'s [=top-level traversable/bounce tracking record=]'s
+    [=bounce tracking record/client bounce detection running=] is false.
+1. Run [=run steps after a timeout=] given <var ignore>milliseconds</var> [=client bounce detection timer period=]
+    and <var ignore>completionSteps</var> [=record stateful bounces for bounce tracking=] given |navigable|.
+1. If |navigable| is not a [=top-level traversable=], then abort these steps.
+1. If |URLs| is empty, then abort these steps.
+1. [=map/For each=] |URL| in |URLs|:
+    1. Let |site| be the result of running [=obtain a site=] given |URL|.
+    1. Let |host| be the |site|'s [=host=].
+    1. [=Insert=] |host| to the |navigable|'s [=top-level traversable/bounce tracking record=]'s [=bounce tracking record/bounce set=].
+
+</div>
+
+<h5 id="bounce-tracking-mitigations-nav-commit-monkey-patch">Commit Navigation Monkey
+Patch</h5>
+
+When the navigation is finalized on a new document, update the [=bounce tracking record/final host=].
+
+Insert the following steps at the end of the <a spec="html">navigate</a> algorithm.
+
+1. Run [=process navigation commit for bounce tracking=] given
```

Comment:
What is the difference between the previous "process navigation end for bounce tracking" and this "process navigation commit for bounce tracking"?

Reply:
Maybe the language is unclear, I'm happy to change it. "End" refers to when the response is received (with a URL list) and "Commit" refers to when the new page is loaded. Chrome has a concept of "committed navigation" (https://chromium.googlesource.com/chromium/src/+/HEAD/docs/navigation.md#Commit), but there is no equivalent cutoff in the W3C spec, so I estimated based on these requirements.

For DIPS purposes, the distinction is important when tracking the final URL. Bounces that match the final URL should be exempt from deletion - so the final URL should ideally represent the final page the user lands on. For instance if the user bounces A->C->B, and unsuccessfully tries to bounce back to C, site C should *not* be exempt from deletion since it might be a tracker.

Let me know if my understanding is correct.

Reply:
Discussed offline - renaming these steps to be more clear: "On response received" and "On document loaded"